File: dev-env.exs
#!/usr/bin/env elixir

Code.require_file("dry_run.ex", __DIR__)

defmodule Setup do
  use DryRun

  def main(args) do
    DryRun.init(args)

    if DryRun.enabled?() do
      IO.puts("DRY RUN - use --apply to execute")
    else
      IO.puts("EXECUTING")
    end

    config_home = System.get_env("XDG_CONFIG_HOME")

    {config_home, message} = if config_home do
      {config_home, "using #{config_home}"}
    else
      home = System.get_env("HOME")
      config_home = Path.join(home, ".config")
      System.put_env("XDG_CONFIG_HOME", config_home)
      {config_home, "XDG_CONFIG_HOME not set\nusing #{config_home}"}
    end

    IO.puts(message)

    copy_files("./env/.config", config_home)

    apply do
      System.cmd("hyprctl", ["reload"], into: IO.stream(:stdio, :line))
    end
  end

  defp copy_files(from, to) do
    IO.puts("\e[35mcopying files: #{from} -> #{to}\e[0m")
    File.cd!(from)

    configs = File.ls!()
    IO.inspect(configs)

    for config <- configs do
      dir = Path.join(to, config)

      IO.puts("removing: rm -rf #{dir}")
      apply do
        File.rm_rf(dir)
      end

      IO.puts("copying: cp -r ./#{config} #{dir}")
      apply do
        File.cp_r("./#{config}", dir)
      end
    end
  end
end

Setup.main(System.argv())

File: dev-env.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Path qw(remove_tree);
use File::Copy::Recursive qw(dircopy);
use File::Copy;
use File::Basename;

my $dry = !grep { $_ eq '--apply' } @ARGV;

if ($dry) {
    print "DRY RUN - use --apply to execute\n";
} else {
    print "EXECUTING\n";
}

my $config_home = $ENV{XDG_CONFIG_HOME};

if ($config_home) {
    print "using $config_home\n";
} else {
    print "XDG_CONFIG_HOME not set\n";
    $config_home = "$ENV{HOME}/.config";
    print "using $config_home\n";
    $ENV{XDG_CONFIG_HOME} = $config_home;
}

copy_files("./env/.config", $config_home);
copy_file("./env/.zshrc", $ENV{HOME});
copy_file("./env/.p10k.zsh", $ENV{HOME});

if ($dry) {
    print "would execute: hyprctl reload\n";
} else {
    system("hyprctl", "reload");
}

sub copy_file {
    my ($from, $to) = @_;

    my $filename = basename($from);
    my $dst = "$to/$filename";

    print "\n\e[35mcopying file: $from -> $dst\e[0m\n";

    if ($dry) {
        print "would execute: copy($from, $dst)\n";
    } else {
        copy($from, $dst) or die "Failed to copy $from: $!\n";
    }
}

sub copy_files {
    my ($from, $to) = @_;

    print "\n\e[35mcopying files: $from -> $to\e[0m\n";

    opendir(my $dh, $from) or die "Cannot open $from: $!";
    my @configs = sort grep { !/^\./ && -d "$from/$_" } readdir($dh);
    closedir($dh);

    print "Configs: @configs\n";

    for my $config (@configs) {
        my $dir = "$to/$config";

        print "removing: rm -rf $dir\n";
        if ($dry) {
            print "would execute: remove_tree($dir)\n";
        } else {
            remove_tree($dir) if -e $dir;
        }

        print "copying: cp -r $from/$config $dir\n";
        if ($dry) {
            print "would execute: dircopy($from/$config, $dir)\n";
        } else {
            dircopy("$from/$config", $dir) or die "Failed to copy $config: $!\n";
        }
    }
}
